ios/Capacitor/Capacitor/KeyPath.swift

```diff
@@ -0,0 +1,62 @@
+import Foundation
+
+public struct KeyPath {
+    var segments: [String]
+    var isEmpty: Bool { return segments.isEmpty }
+    var path: String {
+        return segments.joined(separator: ".")
+    }
+
+    // initializers
+    init(_ string: String) {
+        self.segments = string.components(separatedBy: ".")
+    }
+
+    init(segments: [String]) {
+        self.segments = segments
+    }
+
+    // returns a tuple of the first segment and the remaining key path. result is nil if the key path has no segments.
+    func headAndRemainder() -> (head: String, remainder: KeyPath)? {
+        guard !isEmpty else {
+            return nil
+        }
+        var paths = segments
+        let head = paths.removeFirst()
```

Comment:
I don't know swift very well, does this modify `self.segments`?

Reply:
Short answer: no.

Long answer: collections are value types in Swift (unless you jump through some hoops to make a reference type) so this is only modifying `paths` not `segments`. In addition, KeyPath itself is a struct (which is a copy-on-write value types) so if this method did change `segments` then it would need to be marked as `mutating` or else it wouldn't compile.

Reply:
Thanks!